### PR TITLE
SAR-166 Flatten attribute dictionaries.

### PR DESCRIPTION
**Pull request by kmadisa**
The [fgo](https://github.com/ska-sa/katsim/blob/master/katsim/ska_mpi_dsh_lmc.fgo) file specifies some of the attribute's event properties and those are expected to be written to the database (a file in this case) and/or configure on the device. However upon closer inspection we realised that when starting the device simulator those properties are never really configured when the attributes are created dynamically. This is because of the deeply nested dictionaries created by the fandango file parser for the device attributes.

This PR addresses this issue. With a flat structure, we can be able to set the event properties on the attribute by using the [UserDefaultAttrProp](https://pytango.readthedocs.io/en/stable/server_api/attribute.html?highlight=set_default_pro#userdefaultattrprop) object's methods.

We also removed keys that do not apply to a TANGO device attribute. This will decrease the time it takes to loop through the properties that we need in https://github.com/ska-sa/tango-simlib/blob/250fa66201124db98f7b76ac5e73715eabc835da/tango_simlib/tango_sim_generator.py#L318-L319

Changes were made to the `tango_simlib/tests/test_yaml_builder.py` module since it uses the fandango parser, and we added assert errror messages to make debugging easier.

*Screenshots or code snippets (if appropriate):*

This is the dict structure of the _trackTableBufferMinSizeToLoad_ attribute before
```python
{'alarms': {'delta_t': 'Not specified',
            'delta_val': 'Not specified',
            'extensions': '[]',
            'max_alarm': 'Not specified',
            'max_warning': 'Not specified',
            'min_alarm': 'Not specified',
            'min_warning': 'Not specified'},
 'color': 'Lime',
 'data_format': tango._tango.AttrDataFormat.SCALAR,
 'data_type': tango._tango.CmdArgType.DevLong,
 'database': '10.96.64.100:10000',
 'description': 'Minimum number of coordinates in track buffer to start load table in ACU. \nIt shall be >=5 because ACU is not accepting less than 5 coordinates for proper interpolation.\nIneffective if trackTableBufferingEnabled is false. If true, coordinates are buffered and loaded \nonly above this attribute value.',
 'device': 'mid_dsh_0000/elt/master',
 'dim_x': 1,
 'dim_y': 0,
 'display_unit': 'No display unit',
 'enum_labels': [],
 'events': {'arch_event': {'archive_abs_change': 'Not specified',
                           'archive_period': 'Not specified',
                           'archive_rel_change': 'Not specified',
                           'extensions': '[]'},
            'ch_event': {'abs_change': '1',
                         'extensions': '[]',
                         'rel_change': 'Not specified'},
            'per_event': {'extensions': '[]', 'period': '60000'}},
 'format': '%d',
 'label': 'trackTableBufferMinSizeToLoad',
 'max_alarm': 'Not specified',
 'max_dim_x': 1,
 'max_dim_y': 0,
 'max_value': 'Not specified',
 'min_alarm': 'Not specified',
 'min_value': 'Not specified',
 'model': '10.96.64.100:10000/mid_dsh_0000/elt/master/trackTableBufferMinSizeToLoad',
 'name': 'trackTableBufferMinSizeToLoad',
 'polling': 0,
 'quality': 'ATTR_VALID',
 'standard_unit': 'No standard unit',
 'string': '5',
 'time': 1602497007.603218,
 'unit': '',
 'value': 5,
 'writable': 'READ_WRITE'}

```
and this is the dict structure of the _trackTableBufferMinSizeToLoad_ attribute after our change:
```python
{'abs_change': '1',
 'archive_abs_change': 'Not specified',
 'archive_period': 'Not specified',
 'archive_rel_change': 'Not specified',
 'data_format': tango._tango.AttrDataFormat.SCALAR,
 'data_type': tango._tango.CmdArgType.DevLong,
 'delta_t': 'Not specified',
 'delta_val': 'Not specified',
 'description': 'Minimum number of coordinates in track buffer to start load table in ACU. \nIt shall be >=5 because ACU is not accepting less than 5 coordinates for proper interpolation.\nIneffective if trackTableBufferingEnabled is false. If true, coordinates are buffered and loaded \nonly above this attribute value.',
 'display_unit': 'No display unit',
 'enum_labels': [],
 'event_period': '60000',
 'label': 'trackTableBufferMinSizeToLoad',
 'max_alarm': 'Not specified',
 'max_dim_x': 1,
 'max_dim_y': 0,
 'max_value': 'Not specified',
 'max_warning': 'Not specified',
 'min_alarm': 'Not specified',
 'min_value': 'Not specified',
 'min_warning': 'Not specified',
 'name': 'trackTableBufferMinSizeToLoad',
 'polling': 0,
 'quality': 'ATTR_VALID',
 'rel_change': 'Not specified',
 'standard_unit': 'No standard unit',
 'unit': '',
 'value': 5,
 'writable': 'READ_WRITE'}

```

*Definition of Done Checklist*

- [x] Code meets our [python style guidelines](https://docs.google.com/document/d/1aZoIyR9tz5rCWr2qJKuMTmKp2IzHlFjrCFrpDDHFypM/edit?usp=sharing)?
- [x] Unit tested (coded, passed, included)?
- [x] Requested at least 2 reviewers?
- [x] Commented code, particularly in hard-to-understand areas?
- [x] Made corresponding changes to the documentation (e.g. Python documentation, System Engineering Documentation, version description updates, README file, etc)?

JIRA (SKA): [SAR-166](https://jira.skatelescope.org/browse/SAR-166)

### CODE REVIEW

File: tango_simlib/tests/test_fandango_json_parser.py [hunk 1]
```diff
@@ -37,52 +37,32 @@
...
-            "extensions": "[]",
-        },
-        "ch_event": {
-            "abs_change": "Not specified",
-            "extensions": "[]",
-            "rel_change": "Not specified",
-        },
-        "per_event": {"extensions": "[]", "period": "1000"},
-    },
-    "format": "%s",
```
Comment (ajoubertza) [inline tango_simlib/tests/test_fandango_json_parser.py:71]: Should we remove `format`?  I see it in the attribute config that we can read back. 

```python
dp.get_attribute_config_ex("desiredPointing")
Out[5]: [AttributeInfoEx(alarms = AttributeAlarmInfo(delta_t = 'Not specified', delta_val = 'Not specified', extensions = [], max_alarm = 'Not specified', max_warning = 'Not specified', min_alarm = 'Not specified', min_warning = 'Not specified'), data_format = tango._tango.AttrDataFormat.SPECTRUM, data_type = tango._tango.CmdArgType.DevDouble, description = 'No description', disp_level = tango._tango.DispLevel.OPERATOR, display_unit = 'No display unit', enum_labels = [], events = AttributeEventInfo(arch_event = ArchiveEventInfo(archive_abs_change = 'Not specified', archive_period = 'Not specified', archive_rel_change = 'Not specified', extensions = []), ch_event = ChangeEventInfo(abs_change = 'Not specified', extensions = [], rel_change = 'Not specified'), per_event = PeriodicEventInfo(extensions = [], period = '1000')), extensions = [], format = '%6.2f', label = 'desiredPointing', max_alarm = 'Not specified', max_dim_x = 3, max_dim_y = 0, max_value = 'Not specified', memorized = tango._tango.AttrMemorizedType.NONE, min_alarm = 'Not specified', min_value = 'Not specified', name = 'desiredPointing', root_attr_name = 'Not specified', standard_unit = 'No standard unit', sys_extensions = [], unit = '', writable = tango._tango.AttrWriteType.READ_WRITE, writable_attr_name = 'desiredpointing')]
```

**Comment by amakhaba** [comment]
Ready for Review @kmadisa ?

File: tango_simlib/tests/test_yaml_builder.py [hunk 2]
```diff
@@ -225,7 +230,8 @@ def test_tango_device_builder():
         mocked_device_proxy.attribute_list_query_ex.assert_called()
         mocked_device_proxy.get_property_list.assert_called()
 
-        assert parsed_yaml[0]["meta"]["attributes"] == [
+        attributes = parsed_yaml[0]["meta"]["attributes"]
```
Comment (amakhaba) [inline tango_simlib/tests/test_yaml_builder.py:233]: much more readable thanks for the change.

File: tango_simlib/utilities/fandango_json_parser.py [hunk 3]
```diff
@@ -58,6 +60,37 @@ def parse(self, json_file):
...
+            attribute_config.update(alarms_properties)
+
+    def _remove_unused_device_attributes_keys(self):
+        keys_to_pop = [
+            "alarms",
+            "color",
+            "database",
+            "device",
+            "dim_x",
+            "dim_y",
```
Comment (ajoubertza) [inline tango_simlib/utilities/fandango_json_parser.py:81]: For spectrum and images, do we determine them automatically when setting this initial value for the attribute, or do we just use the maximums?  I'm wondering what would happen if the size of the data in the `value` field didn't match the `max_dim_x` and `max_dim_y`.

File: tango_simlib/utilities/fandango_json_parser.py [hunk 4]
```diff
@@ -58,6 +60,37 @@ def parse(self, json_file):
...
+            "events",
+            "extensions",
+            "format",
+            "model",
+            "string",
+            "time",
+        ]
+        for attribute_config in self._device_attributes.values():
+            # pop out keys not required to configure an attribute
```
Comment (ajoubertza) [inline tango_simlib/utilities/fandango_json_parser.py:90]: This comment seems unnecessary.  The method name already indicates that we are doing this.

**Review by ajoubertza — COMMENTED** [review]
Very nice.  Just a few questions.

File: tango_simlib/tests/test_fandango_json_parser.py (hunk 1 above)
Comment (kmadisa) [reply tango_simlib/tests/test_fandango_json_parser.py:71]: Yes. It is something that tango create automatically depending on the `data_type`. We do not need it to create an attribute dynamically.

File: tango_simlib/utilities/fandango_json_parser.py (hunk 3 above)
Comment (kmadisa) [reply tango_simlib/utilities/fandango_json_parser.py:81]: We only use the maximums. If the size of the data in the `value` field was to be bigger than the `max_dim_x` or `max_dim_y` we would get an error when we try to read the attribute:

```python
DevFailed[
DevError[
    desc = Data size for attribute achievedPointing exceeds given limit
  origin = Attribute::set_value()
  reason = API_AttrOptProp
severity = ERR]

DevError[
    desc = Failed to read_attribute on device mid_dish_0000/elt/master, attribute achievedPointing
  origin = DeviceProxy::read_attribute()
  reason = API_AttributeFailed
severity = ERR]
]
```
which occurs here https://github.com/ska-sa/tango-simlib/blob/250fa66201124db98f7b76ac5e73715eabc835da/tango_simlib/tango_sim_generator.py#L130-L141 in the library for SPECTRUM attributes.

Comment (ajoubertza) [reply tango_simlib/utilities/fandango_json_parser.py:81]: OK, thanks for the detail.

**Review by ajoubertza — APPROVED** [review]
LGTM.

**Pull request merged**